lua/lualine/components/branch.lua

```diff
@@ -1,33 +1,79 @@
-local async = require('lualine.async')
-
 local git_branch
 
-local get_git_branch = async:new({
-  cmd = 'git branch --show-current',
-  on_stdout = function(_, data)
-    if data then
-      git_branch = data:gsub('\n', '')
+local known_git_dirs = {}
+local sep = package.config:sub(1,1)
+
+local function exists(path)
+  local f = io.open(path)
+  if not f then return false, false end
+  local _, _, errno = f:read()
+  f:close()
+  return true, errno == 21 -- is a directory
+end
+
+local function find_git_dir()
+  -- path seperator is not in the end add it
+  local file_path = vim.fn.expand('%:p:h')..sep
+  -- don't show branch for terminal buffers
+  if file_path:match('^term://.*$') then return nil end
+
+  local dir = file_path
+  -- do we need to iterate more than that?
+  local n = 30
+  while dir and n > 0 do
+    n = n - 1
+    -- See if we already know where it is
+    if known_git_dirs[dir] then
```

Comment:
This is another hardcoded variable I don't like but maybe there is no other way of improving performance. On the other hand galaxyline doesn't seem to be using any memorization at first glance, check it out [here](https://github.com/glepnir/galaxyline.nvim/blob/main/lua/galaxyline/provider_vcs.lua). Vimscript plugins like signify also don't use memorization as far as I know

Reply:
This is related to upper one . Just ensures we iterate n times before giving up

Reply:
Won't this create another issue of possibly watching a lot of files.

Also a quick note, this is completely broken for submodules due to memorizing. I personally would prefer to ditch this part of pr if we can test this does not affect performance, I hope it doesn't. If it does we will have to look into an async implementation (other than the old async module)

Reply:
>Won't this create another issue of possibly watching a lot of files?

I'm only watching one file at a time . The watcher is reset before a new file is set to watch.

>Also a quick note, this is completely broken for submodules due to memorizing. 

I haven't tested with submodules I'll update after I've tested.

Reply:
There is no noticeable performance difference between between using or not using branch component on startup so there should be no issue while changing buffers either(without using memorization)

Reply:
Are you sure it's due to memorization . Cause it still doesn't work after commenting out everything with memorization.

Reply:
Submodule bug is not for memorization . It's happening because submodule/.git has a relative path based on submodules dirextory . But that path is interpreted as users curent dir . I'll see how can I fix it.

Reply:
Submodule bug fixed though I donn't like
```        if git_dir:sub(1,2) == '..' then
```
condition to detect relative path in submodules . I'd like something more genaral . Open to suggestions.

Reply:
nevermind, there was an issue with what I was trying to do. Either way, I don't think this small optimization will give us much value and we should remove the memorization part. 
vim-rooter does no memorization at all and runs on WinEnter, BufEnter and I had no issues for 2 years of using it.
Removing it will make the code cleaner while keeping performance marginally worse, I think it's worth it

Reply:
Alright if you are so keen to remove memorization I'll remove it . 

Do you know how full paths start in windows and mac . Like in linux it always starts with '/'

Reply:
on mac they start with / as well, don't remember how they are on windows, why do you need it?

Reply:
the change I suggested below works well for linux, it traverses to root. I will test it if it works the same way on mac an windows later today